### ISSUE TITLE
Metrics get re-registered for builtin-kube-apiserver, aggregator, apiextensions if they are in the same binary

### ISSUE THREAD

**Issue by logicalhan** [kubernetes/kubernetes#75779]
Currently, the same metrics endpoint gets re-registered during the initialization sequence for the builtin apiserver, the aggregator and apiextensions server. While each of these components can exist in isolation, it is not necessarily true that we always want the same metrics for each component; it is furthermore likely undesirable that we register the same metrics over again when the servers coexist in the same binary as eventually we will want to move off of a global metrics registry and this could potentially become an issue. 

/cc @brancz @DirectXMan12 
/sig instrumentation api-machinery

**Comment by brancz** [comment]
I think for various reasons it's good to not use a global metrics registry. I'm all for this. Ideally I'd like to see a few milestones proposed as to how we can slowly refactor the codebase to move to an injected registry as opposed to a global one.

I think a solution that has been vaguely discussed before looked something along the lines of changing all references of `prometheus.DefaultRegisterer`/`prometheus.MustRegister`, to call a package local var, that in turn is just set to the same global registry or use a singleton. Then piece by piece refactor packages to inject the global registry until in the very top packages we only once pass the global registry and we can replace it with a separate one.

Before we start anything like this we should probably have a more fundamental discussion about opencensus as well.

**Comment by sttts** [comment]
How would single-component-in-one-process metrics be published?

**Comment by brancz** [comment]
Still via a `/metrics` endpoint. The point is just that because the registry is global and metrics are registrered via init functions, random etcd metrics are registered onto the global metrics registry for a controller-manager process.

**Comment by sttts** [comment]
> random etcd metrics are registered onto the global metrics registry

Because we have the apiserver dependency. That will go away when https serving moves to component-base.

Who aggregates the right metrics registries per process?

**Comment by brancz** [comment]
There should only be 1 registry that is explicitly created by the component and explicitly passed down.

**Comment by logicalhan** [comment]
> I think for various reasons it's good to not use a global metrics registry. I'm all for this. Ideally I'd like to see a few milestones proposed as to how we can slowly refactor the codebase to move to an injected registry as opposed to a global one.
> 
> I think a solution that has been vaguely discussed before looked something along the lines of changing all references of `prometheus.DefaultRegisterer`/`prometheus.MustRegister`, to call a package local var, that in turn is just set to the same global registry or use a singleton. Then piece by piece refactor packages to inject the global registry until in the very top packages we only once pass the global registry and we can replace it with a separate one.

I really like this idea and I think there are other reasons which would make us want these wrappers; specifically, I believe that having the ability to control the metric registration process is potentially quite powerful.  

> Before we start anything like this we should probably have a more fundamental discussion about opencensus as well.

Definitely. One particularly interesting aspect to think (at least for me) about is opencensus tags when compared to prometheus labels. Would love to discuss that more at the next sig.

**Comment by danielqsj** [comment]
@brancz @logicalhan +1 for opencensus.

**Comment by logicalhan** [comment]
The duplicated metric registration errors reported from #73551 is likely due to this behavior. The errors due to invalid metric names is a separate issue but also easy to fix.
